### PR TITLE
FEATURE: Fetch list of users based on last PR merged

### PR DESCRIPTION
**Pull request by sahsisunny**
## PR Description

This pull request addresses the issue #1429 which aims to fetching a list of users based on their last merged pull request. The purpose of this feature is to identify users who have not contributed to the project in a while, helping us manage our user base more effectively.

### Changes Made

- Added the new query parameter `days` to the API endpoint `users?query=filterBy:UNMERGED_PRS+days:20`
- The endpoint queries the database to retrieve users github_id and fetch the last merged pull request for each user.
- The endpoint then filters the users based on the last merged pull request and the `days` parameter.
- Returns a JSON response containing the list of users and the count of such users.
- Added unit tests to ensure the functionality is working as expected.

### Testing

- I have tested the feature manually and verified that it works as expected.
- I have verified that the response matches the expected results.
- Unit tests have been added and pass successfully.

### Checklist

- [x] Code has been reviewed and meets our coding standards.
- [x] Unit tests have been added and pass successfully.
- [x] Documentation has been updated to include details about the new API endpoint.
- [x] The feature has been tested thoroughly.
- [x] The code has been merged into the main branch.

### Additional Information

Add GitHub Personal Access Token to the .env variable `GITHUB_PERSONAL_ACCESS_TOKEN` to environment variables to avoid rate limiting.

### Test Cases
image
image
image
image

### CODE REVIEW

**Review by Pratiyushkumar — APPROVED** [review]

File: utils/users.js [hunk 1]
```diff
@@ -196,6 +196,25 @@ const getRoleToUpdate = async (userData, newRoles) => {
   return { updateRole: true, newUserRoles };
 };
 
+const parseSearchQuery = (queryString) => {
+  const searchParams = {};
+  const queryParts = queryString.split(" ");
```
Comment (bharati-21) [inline utils/users.js:201]: There is already a function in helpers that splits query string and returns an object. Why are we not re-using that?

File: config/custom-environment-variables.js [hunk 2]
```diff
@@ -29,6 +29,8 @@ module.exports = {
     clientSecret: "GITHUB_CLIENT_SECRET",
   },
 
+  githubAccessToken: "GITHUB_PERSONAL_ACCESS_TOKEN",
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply config/custom-environment-variables.js:32]: it depends on @ankushdharkar

File: controllers/users.js [hunk 3]
```diff
@@ -94,6 +98,34 @@ const getUsers = async (req, res) => {
...
       });
     }
+    if (!transformedQuery?.days && transformedQuery?.filterBy === "unmerged_prs") {
+      return res.boom.badRequest(`Days is required for filterBy ${transformedQuery?.filterBy}`);
+    }
+
+    if (transformedQuery?.filterBy === "unmerged_prs" && transformedQuery?.days) {
+      const days = transformedQuery.days;
+      try {
+        const inDiscordUser = await dataAccess.retrieveUsersWithRole(ROLES.INDISCORD);
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply controllers/users.js:108]: Yes

File: services/githubService.js [hunk 4]
```diff
@@ -260,6 +260,70 @@ const fetchIssues = async () => {
...
+ **/
+const fetchLastMergedPR = async (username) => {
+  try {
+    const baseURL = config.get("githubApi.baseUrl");
+    const issues = "/search/issues";
+    const urlObj = new URL(baseURL);
+    urlObj.pathname = issues;
+    urlObj.searchParams.append("q", `is:pr is:merged author:${username} org:${config.get("githubApi.org")}`);
+    urlObj.searchParams.append("sort", "merged");
+    urlObj.searchParams.append("order", "desc");
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply services/githubService.js:276]: The order parameter in the GitHub API URL controls the global sorting order of search results, impacting the order of all items in the response, not just the pagination.

File: services/githubService.js [hunk 5]
```diff
@@ -260,6 +260,70 @@ const fetchIssues = async () => {
...
+ * @returns {boolean} - True if last PR merged is within the last `days` days else false
+ **/
+const isLastPRMergedWithinDays = async (username, days) => {
+  try {
+    const res = await fetchLastMergedPR(username);
+    const lastPRMergedDate = new Date(res.items[0].pull_request.merged_at);
+    const currentDate = new Date();
+
+    const timeDifferenceInMilliseconds = currentDate - lastPRMergedDate;
+    const timeDifferenceInDays = timeDifferenceInMilliseconds / (1000 * 60 * 60 * 24);
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply services/githubService.js:318]: Which one?

File: utils/users.js (hunk 1 above)
Comment (sahsisunny) [reply utils/users.js:201]: Which function?

File: test/integration/users.test.js [hunk 6]
```diff
@@ -532,6 +532,39 @@ describe("Users", function () {
...
+        .end((err, res) => {
+          if (err) {
+            return done(err);
+          }
+          expect(res).to.have.status(400);
+          expect(res.body).to.be.an("object");
+          expect(res.body.message).to.equal("Days is required for filterBy unmerged_prs");
+          return done();
+        });
+    });
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply test/integration/users.test.js:568]: When filterBy is not passed, it returns all the users, so I think there is no need to write the test for it.

File: test/unit/middlewares/user-validator.test.js [hunk 7]
```diff
@@ -179,4 +179,41 @@ describe("Middleware | Validators | User", function () {
...
+        },
+      };
+      const nextSpy = sinon.spy();
+
+      await getUsers(req, res, nextSpy).catch((err) => {
+        expect(err).to.be.an.instanceOf(Error);
+      });
+      expect(nextSpy.calledOnce).to.be.equal(false);
+    });
+  });
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply test/unit/middlewares/user-validator.test.js:218]: Same as above comment

File: utils/users.js (hunk 1 above)
Comment (sahsisunny) [reply utils/users.js:201]: Okay, got it. However, in this function, I modify the query value by converting it to an integer and lowercase, so we cannot use the getQualifiers function.

File: services/githubService.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (sahsisunny) [reply services/githubService.js:317]: If you are referring to the variables timeDifferenceInMilliseconds and timeDifferenceInDays, we are not using them anywhere, so there is no need to create constants for them.

**Review by Pratiyushkumar — APPROVED** [review]

File: config/default.js [hunk 8]
```diff
@@ -105,4 +105,5 @@ module.exports = {
     APP_ACCESS_KEY: "EVENT_100MS_APP_ACCESS_KEY",
     APP_SECRET: "EVENT_100MS_APP_SECREt",
```
Comment (Achintya-Chatterjee) [inline config/default.js:106]: I think there's a small typo in the Event100ms section; `APP_SECREt` should be corrected to `APP_SECRET.`

**Review by Achintya-Chatterjee — COMMENTED** [review]
There are some duplicate codes for error handling and response in controllers. you can consider creating utility functions to handle common tasks like error handling and responses to reduce duplication and improve code maintenance

**Comment by sahsisunny** [comment]
> There are some duplicate codes for error handling and response in controllers. you can consider creating utility functions to handle common tasks like error handling and responses to reduce duplication and improve code maintenance

Which duplicate code?

File: config/default.js (hunk 8 above)
Comment (sahsisunny) [reply config/default.js:106]: > I think there's a small typo in the Event100ms section; `APP_SECREt` should be corrected to `APP_SECRET.`
😵

Comment (sahsisunny) [reply config/default.js:106]: > I think there's a small typo in the Event100ms section; `APP_SECREt` should be corrected to `APP_SECRET.`

Are you sure it is a typo and I made these changes?

**Comment by shubhamsigdar1** [comment]
Description is written very well and understandable @sahsisunny. Thank you

**Review by bksh05 — COMMENTED** [review]
Nice and clear description with easy-to-understand code
One thing, Did you update the API contract for this change? If yes, can you link it here?
Except that everything else looks good.

File: config/default.js (hunk 8 above)
Comment (sahsisunny) [reply config/default.js:106]: I made these changes?

**Comment by sahsisunny** [comment]
> Nice and clear description with easy-to-understand code One thing, Did you update the API contract for this change? If yes, can you link it here? Except that everything else looks good.

For a GET request, I don't think we need to update the API contracts because we have to change the API query parameter. I have already created an [issue](https://github.com/Real-Dev-Squad/website-backend/issues/1449) ticket for this.

**Review by bksh05 — APPROVED** [review]

**Review by fakhruddinkw — APPROVED** [review]

**Review by Achintya-Chatterjee — APPROVED** [review]
LGTM

**Pull request merged**